sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala

```diff
@@ -534,6 +552,8 @@ final class DataStreamWriter[T] private[sql](ds: Dataset[T]) {
 
   private var foreachWriter: ForeachWriter[T] = null
 
+  private var pythonForeachWriter: PythonForeachWriter = null
+
```

Comment:
I definitely think a discussion is needed here. The reason why I add this is because `foreachWriter` cannot be used here because it's type parameter is the same as the DataStreamWriter's, which is `Row`. But `PythonForeachWriter` extends ForeachWriter[UnsafeRow]:

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonForeachWriter.scala#L33-L34

Therefore I can't just create a `PythonForeachWriter` and call `writer.foreach(pythonForeachWriter)` in `SparkConnectPlanner`.

The original caller is in python:
https://github.com/apache/spark/blob/master/python/pyspark/sql/streaming/readwriter.py#L1309C24-L1315

Maybe that circumvent this check somehow, I'm also very interested if anyone knows why is that

Reply:
I will think about this.. it might be doable since it is ok in PySpark.

Reply:
What if you just try to cast with `asInstanceOf[org.apache.spark.sql.ForeachWriter[org.apache.spark.sql.Row]]` ?

Reply:
> `asInstanceOf[org.apache.spark.sql.ForeachWriter[org.apache.spark.sql.Row]] ?`

Yes, what happens we do this?

Reply:
UnsafeRow does not extend Row. Spark SQL separates the internal representation of Row from public representation, named as InternalRow. UnsafeRow extends InternalRow.